### ISSUE TITLE
Bubble padding is no longer working in version 2.3.1

### ISSUE THREAD

**Issue by funnel20** [chrismiles/CMPopTipView#132]
I just replaced version 2.3.0 by 2.3.1 and compiled my existing project in Xcode 8.3.3.
The bubble is configured with `sidePadding = 10.0`.

When running it on an iPhone 7 with iOS 11.0.3 there is a big difference in padding.

**Version 2.3.0**
![img_4735](https://user-images.githubusercontent.com/2139844/32181655-30e2bd70-bd95-11e7-964c-27f23f75c3fc.JPG)

**Version 2.3.1**
![img_4736](https://user-images.githubusercontent.com/2139844/32181665-380cb998-bd95-11e7-8351-7fabf7490424.JPG)

The padding is no longer present.
Please advise.

**Comment by TingtingAn** [comment]
i am hitting it also.

**Comment by funnel20** [comment]
The Release Notes of 2.3.1 are incomplete by only referring to: 

> Fixes issue in iOS 11 where tooltip would not display when presenting via a UIBarButtonItem.

When I look at the master, 2.3.0 was commit 176, while 2.3.1 is 199. So there are many more changes in 2.3.1.
Commits 185 and 186 refer to a **new property for padding**. 
The change log of commit 186 https://github.com/chrismiles/CMPopTipView/commit/c9ab6ae6a408c8439530e062faacf7006e27982a shows indeed new property `shouldEnforceCustomViewPadding`.

We have reverted back to 2.3.0.
@TingtingAn Can you try with `shouldEnforceCustomViewPadding = TRUE`?

**Comment by TingtingAn** [comment]
i have never to use `initWithCustomView ` also

**Comment by pventura1976** [comment]
Using `sidePadding = 10.0` and `shouldEnforceCustomViewPadding = YES`worked for me in 2.3.1

**Comment by funnel20** [comment]
@pventura1976 Thanks for confirming this.
This seems to be a bug. 
Why is it required when setting property `sidePadding` to explicitly set `shouldEnforceCustomViewPadding`?
At least there is a work-around now, but it doesn't make sense.

**Comment by kleinlieu** [comment]
Hey @FungusHumungus @pventura1976 I'm really sorry for the mistake. Perhaps I should have set `shouldEnforceCustomViewPadding` and `sidePadding` to be on by default in `v2.3.2`. What do you think?

**Comment by funnel20** [comment]
@kleinlieu Thanks for your reaction. Before jumping to a solution; what is the purpose of new property `shouldEnforceCustomViewPadding`?

I would say that if a user wants to have padding, he defines `sidePadding`. 
If he uses property `customView` and wants padding, again he defines `sidePadding`.
For me it's not clear what `shouldEnforceCustomViewPadding` should control.

It seems to be some unnecessary BOOL introduced for the custom view.
But in my opinion it can be completely deleted.
It is now used as IF condition in `- (CGRect)contentFrame` and `- (void)presentPointingAtView:(UIView *)targetView inView:(UIView *)containerView animated:(BOOL)animated`. 
The IF condition can be replaced by:
`if (_ sidePadding > 0) {`

Do you agree?

**Comment by funnel20** [comment]
Another over dimension seems to be by parameters `bubblePaddingX` and `bubblePaddingY`. What is the difference with `sidePadding`?

**Comment by funnel20** [comment]
BTW, `sidePadding` is already set to a default value of 2.0 in `- (id)initWithFrame:(CGRect)frame`.

**Comment by pventura1976** [comment]
I like this proposal

Best.

Pedro Ventura Gómez

**Comment by funnel20** [comment]
As mentioned before, we need to know the functionality first.
I figured it out by playing around with the DEMO project.

**IMPORTANT**
I was able to change `sidePadding` in the default DEMO project, where `shouldEnforceCustomViewPadding` is disabled.
In order to see the proper effect of `cornerRadius`, `bubblePaddingX` and `bubblePaddingY` I had to enable `shouldEnforceCustomViewPadding`. 
Hence I changed the title if this issue from _Property "sidePadding"_ to _Bubble padding_.

There are 5 properties playing a role:

1. `sidePadding`
This is the horizontal offset between the bubble and the view boundary:
![sidepadding](https://user-images.githubusercontent.com/2139844/32979934-621b23dc-cc5e-11e7-9486-55ba0f8fcb35.png)
The default value is 2. In the above screenshot it was set to 10.
2. `cornerRadius`
This will round the corners:
![cornerradius](https://user-images.githubusercontent.com/2139844/32980053-42f867ce-cc60-11e7-8680-e088c0b2ec71.png)
But it also increases the bubble around the text. See the difference with the previous image when increasing the value:
![cornerradius40](https://user-images.githubusercontent.com/2139844/32980060-56b6654a-cc60-11e7-8520-57d1edf8dc8c.png)
The default value is 10.
3. `bubblePaddingX`
![bubblepaddingx](https://user-images.githubusercontent.com/2139844/32980162-fae47502-cc61-11e7-8148-5c6e5a3374f7.png)
This is the horizontal offset between the bubble text and the bubble boundary:
**IMPORTANT**: When `cornerRadius` >0 it's added to `bubblePaddingX`, so the horizontal offset is the total of both properties.
The default value is 0. When this value >0, the horizontal bubble size increases.
4. `bubblePaddingY`
This is the vertical offset between the bubble text and the bubble boundary:
![bubblepaddingy](https://user-images.githubusercontent.com/2139844/32980163-0094153e-cc62-11e7-98c2-f86e4a58b230.png)
**IMPORTANT**: When `cornerRadius` >0 it's added to `bubblePaddingY`, so the vertical offset is the total of both properties.
The default value is 0. When this value >0, the vertical bubble size increases.
5. `shouldEnforceCustomViewPadding`
This seems to be an unnecessary **internal** property which currently causes this issue that bubble padding is not working.

**Comment by funnel20** [comment]
**SOLUTION**
Now that the functionality of all involved properties is clear, the solution seems simple:

1. Remove property `shouldEnforceCustomViewPadding`
2. In `- (CGRect)contentFrame`, the complete IF-ELSE can removed, as the IF statement should _always_ be performed, as it contains all the 3 discussed properties `cornerRadius`, `bubblePaddingX` and `bubblePaddingY`:
```ObjC
- (CGRect)contentFrame {
    CGRect bubbleFrame = [self bubbleFrame];
    
    CGRect contentFrame = CGRectMake(bubbleFrame.origin.x + _cornerRadius + _bubblePaddingX,
                                     bubbleFrame.origin.y + _cornerRadius + _bubblePaddingY,
                                     bubbleFrame.size.width - (_bubblePaddingX*2) - (_cornerRadius*2),
                                     bubbleFrame.size.height - (_bubblePaddingY*2) - (_cornerRadius*2));
    return contentFrame;
}
```
3. Similar in - (void)presentPointingAtView:(UIView *)targetView inView:(UIView *)containerView animated:(BOOL)animated the IF condition should always be performed:
```ObjC
_bubbleSize = CGSizeMake(textSize.width + (_bubblePaddingX*2) + (_cornerRadius*2), textSize.height + (_bubblePaddingY*2) + (_cornerRadius*2));
```

I have verified the DEMO project after these 3 changes and now the bubble padding does properly work without changing anything to the DEMO code.

**Comment by funnel20** [comment]
I have created a pull request with the solution.
Should I make another with more documentation in the header file?

**Comment by funnel20** [comment]
This is resolved as 2.3.2
